### PR TITLE
feat: allow setting samconfig common parameters in [<env>.global.parameters]

### PR DESCRIPTION
**Pull request by wchengru**
#### Which issue(s) does this change fix?
#2382 

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/cli/main.py [hunk 1]
```diff
@@ -58,6 +59,27 @@ def print_info(ctx, param, value):
     ctx.exit()
 
 
+def log_cmdline_info(func):
+    def wrapper(*args, **kwargs):
+        if kwargs["config_file"] and kwargs["config_env"]:
```
Comment (mgrandis) [inline samcli/cli/main.py:64]: Shall we test if `config_file` and `config_env` are in `kwargs` before testing their value?

We risk getting a `KeyError`.

```
if "config_file" in kwargs and "config_env" in kwargs and kwargs["config_file"] and kwargs["config_env"]
```

File: samcli/lib/config/samconfig.py [hunk 2]
```diff
@@ -73,7 +74,14 @@ def get_all(self, cmd_names, section, env=DEFAULT_ENV):
         env = env or DEFAULT_ENV
 
         self._read()
-        return self.document[env][self._to_key(cmd_names)][section]
+        if isinstance(self.document, dict):
+            params = self.document.get(env, {}).get(self._to_key(cmd_names), {}).get(section, {})
```
Comment (mgrandis) [inline samcli/lib/config/samconfig.py:78]: You should put `self.document.get(env, {})` in a variable as you are accessing it 3 times.

File: samcli/lib/config/samconfig.py [hunk 3]
```diff
@@ -106,19 +114,21 @@ def put(self, cmd_names, section, key, value, env=DEFAULT_ENV):
...
-            # in parent layer will override the whole child layer
-            if self.document.get(env, None):
-                if self.document[env].get(self._to_key(cmd_names), None):
-                    if not self.document[env][self._to_key(cmd_names)].get(section, None):
-                        self.document[env][self._to_key(cmd_names)].update({section: {key: value}})
+        # Empty document prepare the initial structure.
+        # self.document is a nested dict, we need to check each layer and add new tables, otherwise duplicated key
+        # in parent layer will override the whole child layer
+        if self.document.get(env, None):
+            if self.document[env].get(self._to_key(cmd_names), None):
```
Comment (mgrandis) [inline samcli/lib/config/samconfig.py:121]: Same here, I would put `self.document[env].get(self._to_key(cmd_names), None)` in a variable

File: samcli/lib/config/samconfig.py [hunk 4]
```diff
@@ -190,6 +200,50 @@ def _write(self):
     def _version(self):
         return self.document.get(VERSION_KEY, None)
 
+    def _deduplicate_global_parameters(self, cmd_names, section, key, value, env=DEFAULT_ENV):
```
Comment (mgrandis) [inline samcli/lib/config/samconfig.py:203]: Are we sure we want to do this?
Potentially erasing customer values.
I think that what is expected is to be able to define global values but also be able to supersede the global value by defining one in a more specific section.

File: samcli/lib/config/samconfig.py (hunk 3 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py]: Rewrite this part to make the if-else branches cleaner.

File: samcli/lib/config/samconfig.py (hunk 2 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py]: done

File: samcli/cli/main.py (hunk 1 above)
Comment (wchengru) [reply samcli/cli/main.py]: Fixed this, thanks!

File: samcli/lib/config/samconfig.py [hunk 5]
```diff
@@ -106,19 +115,26 @@ def put(self, cmd_names, section, key, value, env=DEFAULT_ENV):
...
+        # self.document is a nested dict, we need to check each layer and add new tables, otherwise duplicated key
+        # in parent layer will override the whole child layer
+        cmd_name_key = self._to_key(cmd_names)
+        env_content = self.document.get(env, {})
+        cmd_content = env_content.get(cmd_name_key, {})
+        param_content = cmd_content.get(section, {})
+        if env_content:
+            if cmd_content:
+                if param_content:
+                    self.document[env][cmd_name_key][section].update({key: value})
```
Comment (mgrandis) [inline samcli/lib/config/samconfig.py:128]: You can use `param_content` here.

File: samcli/lib/config/samconfig.py [hunk 6]
```diff
@@ -106,19 +115,26 @@ def put(self, cmd_names, section, key, value, env=DEFAULT_ENV):
...
                 else:
-                    self.document[env].update({self._to_key(cmd_names): {section: {key: value}}})
+                    self.document[env][cmd_name_key].update({section: {key: value}})
```
Comment (mgrandis) [inline samcli/lib/config/samconfig.py:130]: `cmd_content` here

File: samcli/lib/config/samconfig.py [hunk 7]
```diff
@@ -106,19 +115,26 @@ def put(self, cmd_names, section, key, value, env=DEFAULT_ENV):
...
             else:
-                self.document.update({env: {self._to_key(cmd_names): {section: {key: value}}}})
-        # Only update appropriate key value pairs within a section
-        self.document[env][self._to_key(cmd_names)][section].update({key: value})
+                self.document[env].update({cmd_name_key: {section: {key: value}}})
```
Comment (mgrandis) [inline samcli/lib/config/samconfig.py:132]: `env_content` here

File: samcli/lib/config/samconfig.py (hunk 4 above)
Comment (mgrandis) [reply samcli/lib/config/samconfig.py]: I think the global section should set default values that can be overwritten in specific sections.

File: samcli/lib/config/samconfig.py (hunk 5 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py]: updated this one and the other two below.

**Review by mgrandis — APPROVED** [review]

**Review by qingchm — COMMENTED** [review]
Other then the docstring improvement everything looks good to me! Nice to see those helper functions all being tested

File: samcli/lib/config/samconfig.py [hunk 8]
```diff
@@ -190,6 +204,49 @@ def _write(self):
...
+        In case the global parameters contains the same key-value with command parameters,
+        we remove the entry in command parameters
+
+        Parameters
+        ----------
+        cmd_name_key : str
+            key of command name
+
+        section : str
+            Specific section within the command to look into Ex: `parameters`
```
[the remark this answers is not in the document]
Comment (wchengru) [reply samcli/lib/config/samconfig.py]: Updated here and the other two usages. Thanks.

**Review by qingchm — APPROVED** [review]

File: samcli/lib/config/samconfig.py [hunk 9]
```diff
@@ -99,19 +108,24 @@ def put(self, cmd_names, section, key, value, env=DEFAULT_ENV):
...
-                if self.document[env].get(self._to_key(cmd_names), None):
-                    if not self.document[env][self._to_key(cmd_names)].get(section, None):
-                        self.document[env][self._to_key(cmd_names)].update({section: {key: value}})
-                else:
-                    self.document[env].update({self._to_key(cmd_names): {section: {key: value}}})
-            else:
-                self.document.update({env: {self._to_key(cmd_names): {section: {key: value}}}})
-        # Only update appropriate key value pairs within a section
-        self.document[env][self._to_key(cmd_names)][section].update({key: value})
+        # Empty document prepare the initial structure.
```
Comment (hoffa) [inline samcli/lib/config/samconfig.py:111]: Are the `if` changes a refactor?

File: samcli/cli/main.py [hunk 10]
```diff
@@ -58,6 +59,27 @@ def print_info(ctx, param, value):
     ctx.exit()
 
 
+def log_cmdline_info(func):
```
Comment (hoffa) [inline samcli/cli/main.py:62]: Why not simply print `kwargs` as-is? Seems very complex for what's essentially a simulation of what the command could look like.

File: samcli/lib/config/samconfig.py [hunk 11]
```diff
@@ -183,6 +197,46 @@ def _write(self):
...
+            save_global_message = (
+                f'\n\tParameter "{key}={value}" in [{env}.{cmd_name_key}.{section}] is defined as a global '
+                f"parameter [{env}.{DEFAULT_GLOBAL_CMDNAME}.{section}].\n\tThis parameter will be only saved "
+                f"under [{env}.{DEFAULT_GLOBAL_CMDNAME}.{section}] in {self.filepath}."
+            )
+            LOG.info(save_global_message)
+            # Only keep the global parameter
+            items = self.document[env][cmd_name_key][section].copy()
+            items.pop(key)
+            self.document[env][cmd_name_key][section] = items
```
Comment (hoffa) [inline samcli/lib/config/samconfig.py:238]: Why the copy? If this is just deleting `key`, `del self.document[env][cmd_name_key][section][key]` should be enough.

File: samcli/lib/config/samconfig.py (hunk 9 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py:111]: Yes.

File: samcli/lib/config/samconfig.py (hunk 11 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py:238]: No special preference on these two options.

File: samcli/cli/main.py (hunk 10 above)
Comment (wchengru) [reply samcli/cli/main.py:62]: Just want to make the format better.

File: samcli/lib/config/samconfig.py (hunk 11 above)
Comment (hoffa) [reply samcli/lib/config/samconfig.py:238]: Then `del` makes more sense as it's obvious what's going on and avoids unnecessary copy.

File: samcli/cli/main.py (hunk 10 above)
Comment (hoffa) [reply samcli/cli/main.py:62]: I'd argue it's overfitted to a very specific use case (and without docstring `log_cmdline_info` suggests it's generic). If `value` is a `list` or `dict` or `set` then the output will be weird.

File: samcli/lib/config/samconfig.py (hunk 11 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py:238]: updated with del.

File: samcli/cli/main.py (hunk 10 above)
Comment (wchengru) [reply samcli/cli/main.py:62]: Thanks for mentioning this, the docstring was missed. It's added.

**Pull request merged**